File: Sorting/Merge_Sort.c
#include <stdio.h>

void Merge(int[],int,int,int);
void MergeSort(int[],int,int);
void Print(int[],int);

int main()
{

 int num;
 printf("Enter Number Of Elements In Array :\n");
 scanf("%d",&num);
 int arr[num],i;
 printf("Enter Elements Of Array One After Another :\n");
 
for(i=0;i<num;i++)
  scanf("%d",&arr[i]);

 int org_arr[num];
 for(i=0;i<num;i++)
  org_arr[i]=arr[i];               //Storing Original Array in org_array

 printf("Unsorted Array :-\n\n");  //Printing Unsorted Array
 for(i=0;i<num;i++)
  printf("%d\n",org_arr[i]);
  printf("\n");
                  

 MergeSort(arr,0,num-1);          //(num-1) because array indexing goes from 0 to (n-1)
 
 Print(arr,num);

 return 0;
}

void MergeSort(int arr[],int left, int right)
{
 if(left<right)                                   //(left < right) ensures that array has >1 elements                                               
 {
  int mid=(left+right)/2;
  
  MergeSort(arr,left,mid);                       //Recursive calls to reapeatedly divide the array
  MergeSort(arr,mid+1,right);
  Merge(arr,left,mid,right);                          
 }                            
}

void Merge(int arr[],int left, int mid, int right)
{

 int temp[right+1];                                //temp is a temporary array to sort the values
 int i=left,j=mid+1,k=left;
 
 while(i<=mid && j<=right)                         //(i<=mid) and (j<=right) ensures that both subarrays are not empty.
 {

  if(arr[i] < arr[j])
   temp[k++]=arr[i++];
  
  else
   temp[k++]=arr[j++];
 }
 while(i<=mid)                                 //This line puts the remaining elements of 1st subarray to the temporary array.
  temp[k++]=arr[i++];

 while(j<=right)
  temp[k++]=arr[j++];

 for(i=left;i<=right;i++)                     //Copying the temporary array to the required array.
  arr[i]=temp[i];
}


void Print(int arr[], int num)
{
 for(int i=0;i<num;i++)
  printf("%d\n",arr[i]);
}  

File: Searching and Sorting/Merge_Sort.py
def MergeUtility(arr, sorted_arr, l, r):
    if(l < r):

        mid = (l + r) // 2

        MergeUtility(arr, sorted_arr, l, mid)
        MergeUtility(arr, sorted_arr, mid + 1, r)

        Merge(arr, sorted_arr, l, mid, r)

    

def Merge(arr, sorted_arr, l, mid, r):
    i, j, k = l, mid + 1, l

    while(i <= mid and j <= r):
        if(arr[i] <= arr[j]):
            sorted_arr[k] = arr[i]
            i += 1
            k += 1

        else:
            sorted_arr[k] = arr[j]
            k += 1
            j += 1
    
    while(i <= mid):
        sorted_arr[k] = arr[i]
        k += 1
        i += 1

    while(j <= r):
        sorted_arr[k] = arr[j]
        j += 1
        k += 1

    for i in range(l, r + 1):
        arr[i] = sorted_arr[i]

def MergeSort(arr):
    sorted_arr = [0] * len(arr)

    return MergeUtility(arr, sorted_arr, 0, len(arr) - 1)

arr = list(map(int, input("Enter elements of the array: ").split()))

MergeSort(arr)

print("Sorted Array is: ", end = "")
print(*arr)